core/src/kura.rs

```diff
@@ -167,99 +168,179 @@ impl Kura {
 
-        let hash_array = blocks.iter().map(VersionedCommittedBlock::hash).collect();
-
-        let mut guard = self.block_hash_array.lock();
-        *guard = hash_array;
+        {
+            // The none value is set in order to indicate that the blocks exist on disk but
+            // are not yet loaded.
+            let data_array = blocks
+                .iter()
+                .map(VersionedCommittedBlock::hash)
+                .map(|hash| (hash, None))
+                .collect();
+
+            let mut guard = self.block_data.lock();
+            *guard = data_array;
+        }
 
         Ok(blocks)
     }
 
     #[allow(clippy::expect_used, clippy::cognitive_complexity, clippy::panic)]
-    fn kura_recieve_blocks_loop(
+    fn kura_receive_blocks_loop(
         kura: &Kura,
         mut shutdown_receiver: tokio::sync::oneshot::Receiver<()>,
     ) {
-        let mut block_reciever_guard = kura.block_reciever.lock();
-        let mut is_closed = false;
+        let mut written_block_count = kura.block_data.lock().len();
+        let mut should_exit = false;
         loop {
             // If kura receive shutdown then close block channel and write remaining blocks to the storage
             if shutdown_receiver.try_recv().is_ok() {
-                info!("Kura block thread is being shut down");
-                block_reciever_guard.close();
-                is_closed = true;
+                info!("Kura block thread is being shut down. Writing remaining blocks to store.");
+                should_exit = true;
             }
 
-            match block_reciever_guard.try_recv() {
-                // No new blocks would be received
-                Err(_) if is_closed => break,
-                Err(_) => std::thread::sleep(std::time::Duration::from_millis(1)),
-                Ok(new_block) => {
-                    #[cfg(feature = "telemetry")]
-                    if new_block.header().height == 1 {
-                        iroha_logger::telemetry!(msg = iroha_telemetry::msg::SYSTEM_CONNECTED, genesis_hash = %new_block.hash());
-                    }
+            let block_data_guard = kura.block_data.lock();
+            if block_data_guard.len() <= written_block_count {
+                if should_exit {
+                    info!("Kura has written remaining blocks to disk and is shutting down.");
+                    return;
+                }
 
-                    if let Some(path) = kura.block_plain_text_path.as_ref() {
-                        let mut plain_text_file = std::fs::OpenOptions::new()
-                            .create(true)
-                            .append(true)
-                            .open(path)
-                            .expect("Couldn't create file for plain text blocks.");
+                written_block_count = block_data_guard.len();
+                drop(block_data_guard);
+                std::thread::sleep(std::time::Duration::from_millis(1));
```

Comment:
why the sleep here?

Reply:
Because we don't want to spin endlessly and hold valuable locks. When there is nothing to be done, we drop locks, sleep and then reenter the loop again.